### PR TITLE
account/abi: increase error message coverage

### PR DESCRIPTION
**Pull request by dbadoy**
### 2. formatSliceString() - Fixed outputting slice as a zero length array
```
// accounts/abi/error_handling.go, formatSliceString()
// Now
input : {"uint16[]", []uint32{1, 2, 3}}
error : {"abi: cannot use []uint32 as type [0]uint16 as argument"}
```
I think this part need fix the format like below.

```
// Fixed
input : {"uint16[]", []uint32{1, 2, 3}}
error : {"abi: cannot use []uint32 as type []uint16 as argument"}
```

### 3. NewType() - Catch zero length array
Added logic that catch the case array of length zero.

### CODE REVIEW

File: accounts/abi/error_handling.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ var (
 // formatSliceString formats the reflection kind with the given slice size
 // and returns a formatted string representation.
 func formatSliceString(kind reflect.Kind, sliceSize int) string {
-	if sliceSize == -1 {
+	if sliceSize < 1 {
```
Comment (dbadoy) [inline accounts/abi/error_handling.go:32]: This method outputs uint[0] as []uint. But zero length array will not input.

**Comment by rjl493456442** [comment]
Hi, thanks for the PR!

I just checked the solidity document, it said 

> <type>[M]: a fixed-length array of M elements, M >= 0, of the given type.
> Note While this ABI specification can express fixed-length arrays with zero elements, they’re not supported by the compiler.

The zero size array is supported by ABI specification itself, but not solidity. Perhaps some other languages(e.g. vyper) can support zero-size array?

**Comment by dbadoy** [comment]
I can't found that documents, only tested it.
```
# vyper compiler version: 0.2 ~
var1: uint256[0]
var2: address[0]

ArrayIndexException: Subscript must be greater than 0


# vyper compiler version: ~ 0.1.x
var1: uint256[0]
var2: address[0]

vyper.exceptions.InvalidTypeException: line 1:14 Arrays / ByteArrays must have a positive integral number of elements
```

Actually, I'm reconsidering whether It need to put a constraint on ethereum client.
This PR's main purpose is `Fixed outputting slice as a zero length array`. 
And It's can resolve below codes,

```go
func NewType(t string, internalType string, components []ArgumentMarshaling) (typ Type, err error) {
    .
    .
}
```

File: accounts/abi/type_test.go [hunk 2]
```diff
@@ -206,12 +206,15 @@ func TestTypeCheck(t *testing.T) {
...
 		{"uint16[]", nil, []uint16{1, 2, 3}, ""},
 		{"uint16[]", nil, [3]uint16{1, 2, 3}, ""},
-		{"uint16[]", nil, []uint32{1, 2, 3}, "abi: cannot use []uint32 as type [0]uint16 as argument"},
+		{"uint16[]", nil, []uint32{1, 2, 3}, "abi: cannot use []uint32 as type []uint16 as argument"},
+		{"uint16[0]", nil, [0]uint32{}, "abi: cannot use [0]uint32 as type [0]uint16 as argument"},
 		{"uint16[3]", nil, [3]uint32{1, 2, 3}, "abi: cannot use [3]uint32 as type [3]uint16 as argument"},
 		{"uint16[3]", nil, [4]uint16{1, 2, 3}, "abi: cannot use [4]uint16 as type [3]uint16 as argument"},
 		{"uint16[3]", nil, []uint16{1, 2, 3}, ""},
-		{"uint16[3]", nil, []uint16{1, 2, 3, 4}, "abi: cannot use [4]uint16 as type [3]uint16 as argument"},
+		{"uint16[3]", nil, []uint16{1, 2, 3, 4}, "abi: cannot use []uint16 as type [3]uint16 as argument"},
```
[the remark this answers is not in the document]
Comment (dbadoy) [reply accounts/abi/type_test.go:214]: Yes... That's not make sense.
I think it need split the logging error about same type, difference length.
In now, If input `uint16[3], []uint16{1,2,3,4}`, the errors is `abi: cannot use [4]uint16 as type [3]uint16 as argument`. But really  `[]uint16{1,2,3,4} == [4]uint16` is true?

IMHO It's the better logs.
`abi: cannot use %v of length %d as type %v`
`abi: cannot use []uint16 of length 4 as type [3]uint16`

What do you think?

Example
```go
func sliceTypeCheck(t Type, val reflect.Value) error {
	if val.Kind() != reflect.Slice && val.Kind() != reflect.Array {
		return typeErr(formatSliceString(t.GetType().Kind(), t.Size), val.Type())
	}

	if t.T == ArrayTy && val.Len() != t.Size {
                 // Added
		if val.Kind() == reflect.Slice {
			return tempErr(formatSliceString(t.Elem.GetType().Kind(), t.Size), val.Type(), val.Len())
		}
		return typeErr(formatSliceString(t.Elem.GetType().Kind(), t.Size), formatSliceString(val.Type().Elem().Kind(), val.Len()))
	}

	if t.Elem.T == SliceTy || t.Elem.T == ArrayTy {
		if val.Len() > 0 {
			return sliceTypeCheck(*t.Elem, val.Index(0))
		}
	}

	if val.Type().Elem().Kind() != t.Elem.GetType().Kind() {
		return typeErr(formatSliceString(t.Elem.GetType().Kind(), t.Size), val.Type())
	}

	return nil
}

// Added
func tempErr(expected, got interface{}, len int) error {
	return fmt.Errorf("abi: cannot use %v of length %d as type %v", got, len, expected)
}
```

**Comment by dbadoy** [comment]
This PR seems no reason to accept. I'll close this!

**Pull request closed without merging**